### PR TITLE
ZFP carrier: update find package and code to the latest library version

### PR DESCRIPTION
**Pull request by barbalberto**
From zfp library version 0.5.1 the library providers start using CMake as a build system.
Some function API changed, so our find package and code has been updated.

### CODE REVIEW

File: cmake/ycm-0.3.1/find-modules/FindZFP.cmake [hunk 1]
```diff
@@ -27,27 +27,28 @@
 include(FindPackageHandleStandardArgs)
 include(SelectLibraryConfigurations)
 
-find_path(ZFP_INCLUDE_DIR
+find_path(ZFP_INCLUDE_DIRS
           NAMES zfp.h
-          PATHS $ENV{ZFP_ROOT}/inc
+          PATHS $ENV{ZFP_DIR}/include
```
Comment (drdanz) [inline cmake/ycm-0.3.1/find-modules/FindZFP.cmake]: Why not including both paths (`inc` and `include`) here? was it broken in the first place?

File: cmake/ycm-0.3.1/find-modules/FindZFP.cmake [hunk 2]
```diff
@@ -27,27 +27,28 @@
 include(FindPackageHandleStandardArgs)
 include(SelectLibraryConfigurations)
 
-find_path(ZFP_INCLUDE_DIR
+find_path(ZFP_INCLUDE_DIRS
```
Comment (drdanz) [inline cmake/ycm-0.3.1/find-modules/FindZFP.cmake]: This change is wrong please revert it.

File: src/carriers/zfp_portmonitor/CMakeLists.txt [hunk 3]
```diff
@@ -8,24 +8,34 @@ yarp_prepare_plugin(zfp TYPE ZfpMonitorObject
...
-               COMPONENT runtime
-               LIBRARY DESTINATION ${YARP_DYNAMIC_PLUGINS_INSTALL_DIR}
-               ARCHIVE DESTINATION ${YARP_STATIC_PLUGINS_INSTALL_DIR})
-  yarp_install(FILES zfp.ini
-               COMPONENT runtime
-               DESTINATION ${YARP_PLUGIN_MANIFESTS_INSTALL_DIR})
-
-  set_property(TARGET yarp_pm_zfp PROPERTY FOLDER "Plugins/Port Monitor")
+
+    if(ZFP_FOUND)
```
Comment (drdanz) [inline src/carriers/zfp_portmonitor/CMakeLists.txt]: Do not put the if here, instead add it to the `DEPENDS` list above.

**Comment by drdanz** [comment]
If I understand correctly, the latest version has the CMake config files, am I wrong? In this case you should be able to do something like this at the beginning:

```cmake
include(StandardFindModule)
standard_find_module(ZFP SKIP_PKG_CONFIG)

if(NOT ZFP_FOUND)
# ...
endif()
```

File: cmake/ycm-0.3.1/find-modules/FindZFP.cmake (hunk 1 above)
Comment (barbalberto) [inline cmake/ycm-0.3.1/find-modules/FindZFP.cmake]: The old version of the library was using `inc`, the new version uses `include`.
I prefer to avoid support older version with different folder structure if not strictly required.

**Comment by barbalberto** [comment]
In the latest version they uses cmake as a build system, but they do not provide config and version files

File: src/carriers/zfp_portmonitor/CMakeLists.txt (hunk 3 above)
Comment (barbalberto) [inline src/carriers/zfp_portmonitor/CMakeLists.txt]: I don't like this option because if the `zfp` library is not found on the system, then the option for creating this plugin is not created and the plugin is not visible among the cmake options.
Therefore people does not even know that such a plugin exists, and there is no information about what to do to enable it.

Vice versa, the way I did the flag is visible, so people do know the plugin is available. If someone tries to use it but the library is not found, the error message will tell what to do.

This is far more user friendly.

File: cmake/ycm-0.3.1/find-modules/FindZFP.cmake (hunk 2 above)
Comment (barbalberto) [inline cmake/ycm-0.3.1/find-modules/FindZFP.cmake]: Can you please better explain why is it wrong?
Before it was:
```
find_path(ZFP_INCLUDE_DIR
          NAMES zfp.h
          PATHS $ENV{ZFP_ROOT}/inc
          DOC "ZFP include directory")
...
set(ZFP_INCLUDE_DIRS ${ZFP_INCLUDE_DIR})
```
The last line looks like an addictional line with no actual effect to me.

Now it is:
```
find_path(ZFP_INCLUDE_DIRS
          NAMES zfp.h
          PATHS $ENV{ZFP_ROOT}/inc
          DOC "ZFP include directory")
```
The result should be the same imho. What am I missing?

File: src/carriers/zfp_portmonitor/CMakeLists.txt (hunk 3 above)
Comment (drdanz) [inline src/carriers/zfp_portmonitor/CMakeLists.txt]: This is how all the other YARP plugins do. It makes the CMake file a lot shorter and easier to read.
They are not displayed as option that can be enabled but the output tells you all you need to know...

File: cmake/ycm-0.3.1/find-modules/FindZFP.cmake (hunk 1 above)
Comment (drdanz) [inline cmake/ycm-0.3.1/find-modules/FindZFP.cmake]: If there is no reason for not supporting both, I don't see why not. Anyway in this way, if a user has the old version, cmake will not find the library but the user will not have a feedback on the reason...

File: cmake/ycm-0.3.1/find-modules/FindZFP.cmake (hunk 2 above)
Comment (drdanz) [inline cmake/ycm-0.3.1/find-modules/FindZFP.cmake]: DIR = cached & only one directory. This is the variable you will set.
DIRS = might be more than one directory and not cached. This is the variable you want to use.
In this way you are caching the wrong variable

File: cmake/ycm-0.3.1/find-modules/FindZFP.cmake (hunk 1 above)
Comment (barbalberto) [inline cmake/ycm-0.3.1/find-modules/FindZFP.cmake]: Well, they changed build system (so header/library position in the filesystem), folder names and also API (so changes in the code are required).
They do not provide versioning, so even if we find all the pieces, we shall have some ifdef in the code depending on the version, but I don't see an easy way to detect the version of the library.

Since this is gonna be the actual first release of this plugin, I believe that starting with a clean situation is better. CMake will tell the required version of the library is 0.5.1, so it is well clear to the user where is the problem.
In case the user has an older version he/she just need to update it.

Comment (drdanz) [inline cmake/ycm-0.3.1/find-modules/FindZFP.cmake]: I see... ok I don't really care about this change!
We should check if the library contains some file where we could read the version though, so that we can perform a proper check `find_package(ZFP 0.5.1)

File: src/carriers/zfp_portmonitor/CMakeLists.txt (hunk 3 above)
Comment (barbalberto) [inline src/carriers/zfp_portmonitor/CMakeLists.txt]: I like the idea of grouping devices together and showing/hiding them as a whole.
I changed it because right now in `devel`, when the library is not  found the output says:
```
 CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
 Please set them or make sure they are set and tested correctly in the CMake files:
 ZFP_INCLUDE_DIR (ADVANCED)
    used as include directory in directory /usr/local/src/robot/ROBOT_CODE/01_yarp/yarp/src/carriers/zfp_portmonitor
    used as include directory in directory /usr/local/src/robot/ROBOT_CODE/01_yarp/yarp/src/carriers/zfp_portmonitor
    used as include directory in directory /usr/local/src/robot/ROBOT_CODE/01_yarp/yarp/src/carriers/zfp_portmonitor
    used as include directory in directory /usr/local/src/robot/ROBOT_CODE/01_yarp/yarp/src/carriers/zfp_portmonitor
    used as include directory in directory /usr/local/src/robot/ROBOT_CODE/01_yarp/yarp/src/carriers/zfp_portmonitor
    used as include directory in directory /usr/local/src/robot/ROBOT_CODE/01_yarp/yarp/src/carriers/zfp_portmonitor
    used as include directory in directory /usr/local/src/robot/ROBOT_CODE/01_yarp/yarp/src/carriers/zfp_portmonitor
    used as include directory in directory /usr/local/src/robot/ROBOT_CODE/01_yarp/yarp/src/carriers/zfp_portmonitor
    used as include directory in directory /usr/local/src/robot/ROBOT_CODE/01_yarp/yarp/src/carriers/zfp_portmonitor
    used as include directory in directory /usr/local/src/robot/ROBOT_CODE/01_yarp/yarp/src/carriers/zfp_portmonitor
```
Which is confusing and not meaningful.

Anyway, it is not true that using `zfp` as a dependency
>   the output tells you all you need to know...

What actually happens is that the plugin will be hidden and, if you really look carefully the cmake will only write
`--  --- plugin yarpcar_zfp: disabled (dependencies unsatisfied)`
without telling what is missing and what has to be done to satisfy this dependency.

With the patch proposed the error message will be:
```
ZFP carrier: ZFP library not found. 
 Please download the 'ZFP' library from 'http://computation.llnl.gov/projects/floating-point-compression',
 build it using cmake and install it. Set the environment variable 'ZFP_DIR' to the installation folder. 
```
Which is more clear and useful.

I like the idea of hiding some flags depending of options to shorten the cmake, like using `CREATE_OPTIONAL_CARRIERS` as a selector, but I do believe that 'logical selection' shall hide things while missing dependencies shall not, they have to raise clear error to the user instead hiding stuff under the carpet.

All the devices I know uses works the same as this plugin, and historically this is always been the way they worked and how the user expect them to be.
All the devices always appears in the list (once the `complie_XXX` is enabled) and when the single device has been selected, dependency errors are thrown.

Comment (drdanz) [inline src/carriers/zfp_portmonitor/CMakeLists.txt]: If you set `YarpPlugin_VERBOSE` the output is more explicit (but I think there isn't an option for enabling that from the GUI)

Anyway this is disappearing, the output you want to give is mostly the same produced by FeatureSummary, and I already have a branch that does it for all the branches and will be merged before the release, so please revert this change as well to save me from looking at this again when I merge the patch.

File: cmake/ycm-0.3.1/find-modules/FindZFP.cmake [hunk 4]
```diff
@@ -29,25 +29,27 @@ include(SelectLibraryConfigurations)
 
 find_path(ZFP_INCLUDE_DIR
           NAMES zfp.h
-          PATHS $ENV{ZFP_ROOT}/inc
+          PATHS $ENV{ZFP_DIR}/include
```
Comment (drdanz) [inline cmake/ycm-0.3.1/find-modules/FindZFP.cmake]: I just noticed this...
Why did you change the name of the environment variable? If there is not a good reason from the upstream changes, the convention is:
ROOT = root of the library installation
DIR = directory containing cmake files

File: cmake/ycm-0.3.1/find-modules/FindZFP.cmake [hunk 5]
```diff
@@ -29,25 +29,27 @@ include(SelectLibraryConfigurations)
...
           DOC "ZFP include directory")
+
```
Comment (drdanz) [inline cmake/ycm-0.3.1/find-modules/FindZFP.cmake]: Please remove this unneeded change

File: cmake/ycm-0.3.1/find-modules/FindZFP.cmake [hunk 6]
```diff
@@ -29,25 +29,27 @@ include(SelectLibraryConfigurations)
...
 find_library(ZFP_LIBRARY_RELEASE
              NAMES ZFP zfp
-             PATHS $ENV{ZFP_ROOT}/lib
+             PATHS $ENV{ZFP_DIR}/lib
              DOC "ZFP library file (release version)")
+
```
Comment (drdanz) [inline cmake/ycm-0.3.1/find-modules/FindZFP.cmake]: same

File: cmake/ycm-0.3.1/find-modules/FindZFP.cmake (hunk 4 above)
Comment (barbalberto) [inline cmake/ycm-0.3.1/find-modules/FindZFP.cmake]: Yes, DIR is the correct one.

Comment (drdanz) [inline cmake/ycm-0.3.1/find-modules/FindZFP.cmake]: No it's not, ROOT is the correct one, I don't see a `ZFPConfig.cmake` file in `ZFP_DIR`

File: src/carriers/zfp_portmonitor/CMakeLists.txt (hunk 3 above)
Comment (barbalberto) [inline src/carriers/zfp_portmonitor/CMakeLists.txt]: I did the revert, hoping this `FeatureSummary` will improve the situation.
Anyway for the future:
- Request changes based on something on a private branch and not known isn't a good practise
- Change an established status quo shall be discussed beforehand

File: cmake/ycm-0.3.1/find-modules/FindZFP.cmake (hunk 4 above)
Comment (traversaro) [inline cmake/ycm-0.3.1/find-modules/FindZFP.cmake]: Related :  https://gitlab.kitware.com/cmake/cmake/merge_requests/796 .

File: src/carriers/zfp_portmonitor/CMakeLists.txt [hunk 7]
```diff
@@ -5,27 +5,29 @@
...
-  set(CMAKE_INCLUDE_CURRENT_DIR ON)
-  get_property(YARP_OS_INCLUDE_DIRS TARGET YARP_OS PROPERTY INCLUDE_DIRS)
-  get_property(YARP_sig_INCLUDE_DIRS TARGET YARP_sig PROPERTY INCLUDE_DIRS)
-  include_directories(${YARP_OS_INCLUDE_DIRS} ${YARP_sig_INCLUDE_DIRS})
-  include_directories(SYSTEM ${ZFP_INCLUDE_DIRS})
 
-  yarp_add_plugin(yarp_pm_zfp zfpPortmonitor.cpp
-                              zfpPortmonitor.h)
-  target_link_libraries(yarp_pm_zfp YARP_OS YARP_sig ${ZFP_LIBRARIES})
+    set(CMAKE_INCLUDE_CURRENT_DIR ON)
```
Comment (drdanz) [inline src/carriers/zfp_portmonitor/CMakeLists.txt]: please keep the indentation at 2 spaces in CMake files

**Comment by drdanz** [comment]
I rebased, reverted unneeded changes and added a `ZFP_VERSION` variable, and checked that ZFP is >= 0.5.1.
Now it looks good to me...
